### PR TITLE
[FLINK-31293][runtime] LocalBufferPool request overdraft buffer only when no available buffer and pool size is reached

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*In our current implementation, once there is no available buffer, overdraft buffers can. be requested, which does not meet our requirements for the name the `overdraft`. In fact, we expect only request additional buffers when the pool size is reached.*

*Another key point is the definition of LocalBufferPool's `available` status. Actually, before overdraft buffer was introduced, its definition was very clear: 
There is at least one `availableMemorySegment` and no subpartitions has reached `maxBuffersPerChannel`. 
IMO, Introducing the overdraft mechanism should not break this protocol. Otherwise, it may cause some very hidden bugs. 
It should be noted that even if we only allow request overdraft buffers after reaches the `poolSize`, there may still be a situation where both `availableMemorySegment` and overdraft buffer are all not zero, and this state should obviously be defined as available.*

## Brief change log

  - *Add missing @GuardedBy annotation for LocalBufferPool.*
  - *LocalBufferPool request overdraft buffer only when no available buffer and pool size is reached.*

## Verifying this change

Manually test this change by running TPC-DS and added unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by reswqa** [comment]
@1996fanrui and @wsry, would you mind taking a look at this? This pr currently can be reviewed, but not the final status. Because there may be some tests to be added and migrated to junit5.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:399]: I'm not sure whether `availableMemorySegments.poll();` should be executed when `availableMemorySegments.isEmpty()`. 

1. How about call it just when `!availableMemorySegments.isEmpty()`?
2. We should add some comments for `requestOverdraftMemorySegmentFromGlobal`, explain when to request the overdraft buffer.

```suggestion
            if (!availableMemorySegments.isEmpty()) {
                segment = availableMemorySegments.poll();
            } else if (isRequestedSizeReached()) {
                // Only when the buffer request reaches the upper limit, requests an overdraft buffer
                segment = requestOverdraftMemorySegmentFromGlobal();
            }
```

Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:399]: IIUC, poll buffer from empty `availableMemorySegments ` is not necessary as we guarded this field by lock.

**Comment by reswqa** [comment]
Thanks @1996fanrui for the review, I have updated this as your comment, PTAL.

**Review by 1996fanrui — APPROVED** [review]
Thanks for the fix, LGTM.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (1996fanrui) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:338]: There may be bug here.

I change the 10 to 7, it should be unavailable, however, it's available. 

`bufferPool.setNumBuffers(7);` requested new buffer from global buffer pool to the `availableMemorySegments`.

Call stack is as follow :

```
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.requestMemorySegmentFromGlobal(LocalBufferPool.java:442)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.checkAvailability(LocalBufferPool.java:551)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.checkAndUpdateAvailability(LocalBufferPool.java:524)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.setNumBuffers(LocalBufferPool.java:682)
	  - locked <0xa5d> (a java.util.ArrayDeque)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPoolTest.testIncreasePoolSize(LocalBufferPoolTest.java:331)
```

I think `isRequestedSizeReached` should be updated.


```
    private boolean isRequestedSizeReached() {
        return numberOfRequestedMemorySegments + numberOfRequestedOverdraftMemorySegments
                >= currentPoolSize;
    }
```

WDYT?

BYW, we should add more tests here, such as : newPoolSize=7.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (1996fanrui) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:314]: testDecreasePoolSize is similar as well.

**Review by 1996fanrui — CHANGES_REQUESTED** [review]
Hi @reswqa , thanks for your update, I have left some comments, please take a look in your free time.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (1996fanrui) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:338]: Or, should we reset the `numberOfRequestedMemorySegments` and `numberOfRequestedOverdraftMemorySegments` inside of the `setNumBuffers`?

If `currentPoolSize` is increased, we should decrease the `numberOfRequestedOverdraftMemorySegments` and increase `numberOfRequestedMemorySegments`.

I'm not sure which solution is better. And I prefer reset them inside of  the `setNumBuffers`, because it's more clear, and other solutions may have bug when `returnMemorySegment`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java]: Good suggestion.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java]: Good catch !!! I think the second one is more better. If pool size becomes larger, some overdraft buffers need convert to requested buffers.

**Comment by reswqa** [comment]
Thanks @1996fanrui for the review and catch the bug, I have updated this according to your comments, please take a look again.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:674]: For the case that the pool size becomes smaller, I prefer not to do special handling, because we already allow the case that `numRequestedBuffers` exceed the `pool size` after `setNumbers`. WDYT?

Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:674]: I see it can work now due to we return overdraft buffer first, right?

However I think convert `numberOfRequestedMemorySegments` to `numberOfRequestedOverdraftMemorySegments ` is more clear, and the semantic of these fields are more explicit.

I'm not sure which one is better. WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: Why the `numberOfRequestedOverdraftMemorySegments  == 0` can be removed? I think there is a constraint here: The `availableMemorySegments` is always empty when `numberOfRequestedOverdraftMemorySegments > 0`. 

Sometimes this constraint does not hold before this PR, and we want to hold the constraint in the future, right?

If yes, could you add some comments for `numberOfRequestedOverdraftMemorySegments`? It's helpful for other developers to understand the constraint.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (1996fanrui) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:319]: I prefer change the `testIncreasePoolSize` to the `testIncreasePoolSizeExceedTotalBuffers`. And the `testIncreasePoolSizeExceedTotalBuffers` and `testIncreasePoolSizeNotExceedTotalBuffers` as the normal method instead of test method. 

And create a new `testIncreasePoolSize()` method as the test method, it calls `testIncreasePoolSizeExceedTotalBuffers` and `testIncreasePoolSizeNotExceedTotalBuffers`. We can define these parameters inside of the `testIncreasePoolSize()`, such as: `largePoolSize`, `smallPoolSize` and `maxOverdraftBuffers`.

Why do I prefer this? We can test more cases for increasing pool size, especially some boundary conditions, such as `largePoolSize = 7` and `largePoolSize = 8`.  The 6 and 10 are not boundary values, and some corner bugs cannot be tested. 

WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:674]: The reason why I didn't do this convert here is that we have an upper limit on the number of overdraft buffers. If this conversion is allowed, and the `pool size` changes very small, the `numberOfRequestedOverdraftMemorySegments ` will exceed its upper limit.
Of course, we can convert it to the upper limit at most, but this will make the logic a little complicated. It seems that there is no benefit to introduce this mechanism. WDTY?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: Before we allows convert  `numberOfRequestedOverdraftMemorySegments` to `numberOfRequestedMemorySegments `, there still be a situation where both `availableMemorySegment` and `overdraft buffer` are all not zero, and this state should obviously be defined as available. But after that, I'm not sure whether this situation still exists. I'm a little worried about the potential bug of multithreading if we 
 add `numberOfRequestedOverdraftMemorySegments  == 0` back to `shouldBeAvailable`. 
For me, the key point is that if we think "`The availableMemorySegments` is always empty when `numberOfRequestedOverdraftMemorySegments` != 0." is tenable, does `!availableMemorySegments.isEmpty()` already include `numberOfRequestedOverdraftMemorySegments == 0`? Even if this will not introduce bug, why should we impose useless constraints? 
Actually, before overdraft buffer was introduced, the definition of `available` was very clear:
There is at least one `availableMemorySegment` and no subpartitions has reached `maxBuffersPerChannel`.
IMO, Introducing the overdraft mechanism should not break this protocol, overdraft buffer should not affect the judgment of availability.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:319]: Agree with you, fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: > Actually, before overdraft buffer was introduced, the definition of available was very clear:
There is at least one availableMemorySegment and no subpartitions has reached maxBuffersPerChannel.
IMO, Introducing the overdraft mechanism should not break this protocol, overdraft buffer should not affect the judgment of availability.

Sorry, overdraft buffer should affect the judgment of availability.

As we discussed before: overdraft buffer is just used when requested buffer reached the upper limit(pool size). In the other word: overdraft buffer should be used after LocalBufferPool is unavailable. 

And why the name is `overdraft`? It temporarily uses some extra buffers outside the LocalBufferPool. From the semantics of overdraft, if `numberOfRequestedOverdraftMemorySegments > 0`, then LocalBufferPool must be unavailable. That's why I add it here.

Why you want to remove it? I guess it has bug before, that is, the overdraft buffer is used when the `requested buffer` does not reach the upper limit, and you have fixed it in this PR.

```
            if (!availableMemorySegments.isEmpty()) {
                segment = availableMemorySegments.poll();
            } else if (isRequestedSizeReached()) {
                // Only when the buffer request reaches the upper limit(i.e. current pool size),
                // requests an overdraft buffer.
                segment = requestOverdraftMemorySegmentFromGlobal();
            }
```

> For me, the key point is that if we think "The availableMemorySegments is always empty when numberOfRequestedOverdraftMemorySegments != 0." is tenable.

If it's tenable now and future, we can remove the ` && numberOfRequestedOverdraftMemorySegments == 0` here, if not, it cannot be removed.

Based on your feedback, I prefer keep it, because the root cause is: overdraft is misused in some cases, and you have fixed it.

If there is other bug that misuse overdraft buffer, we should fix the bug to ensure overdraft buffer is used correctly instead of mark LocalBufferPool is available. Marking the LocalBufferPool is available directly may cause other unexpected bugs.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (1996fanrui) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:319]: > We can test more cases for increasing pool size, especially some boundary conditions, such as largePoolSize = 7 and largePoolSize = 8. The 6 and 10 are not boundary values, and some corner bugs cannot be tested.

Could you add `largePoolSize = 7 and largePoolSize = 8` in this test?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: > Before we allows convert numberOfRequestedOverdraftMemorySegments to numberOfRequestedMemorySegments , there still be a situation where both availableMemorySegment and overdraft buffer are all not zero.

Just as what I said, only after we supports convert `numberOfRequestedOverdraftMemorySegments` to `numberOfRequestedMemorySegments`, then can we avoid having both `available buffers` and `overdraft buffers`. 
It is not enough to just modify the following code as there are still concurrency problems.
```
            if (!availableMemorySegments.isEmpty()) {
                segment = availableMemorySegments.poll();
            } else if (isRequestedSizeReached()) {
                // Only when the buffer request reaches the upper limit(i.e. current pool size),
                // requests an overdraft buffer.
                segment = requestOverdraftMemorySegmentFromGlobal();
            }
```
Of course, we have allowed this conversion now, but I'm not very sure whether there are other concurrency problems even though after this fix.
In any case, your reason is also make sense and acceptable to me. But for the sake of safety, perhaps we should listen to the opinions of @wsry who knows more about the implementation of `LocalBufferPool` than me.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java
Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java:319]: Sure, I added a test for `largePoolSize = 7`.
As for `largePoolSize=8`, I think it is no different from the current situation of `largePoolSize = 10`. It also exceeds the total number of buffers but does not reach `maxMemorySegments`. But I changed the previous test to `largePoolSize = 8`, and added test for the case of total number of buffers reached `maxMemorySegments`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: Hi @pnowojski @akalash , could you help take a look in your free time, thanks

Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: I actually mostly agree that `if (isRequestedSizeReached())` condition and `while` inside `setNumBuffers` fix the bug of this ticket. It is difficult to say something about removing `numberOfRequestedOverdraftMemorySegments == 0`. On one hand, it is better doesn't change anything and improve it in another PR(if we still think it makes sense) since it doesn't relate to this bug, on the other hand, in the worst case(if we still have a bug) removing this condition just allows us to allocate 'ordinary' buffer while we have already overdraft buffers which is not so good but still better than have a deadlock. Let me think about it a little more. But anyway if we will decide to remove it I think it makes sense to do it in a separate commit just to understand where is the fix and where is the prophylactic/improvement

In general, I would say that it seems we need to improve/refactor this class a little(maybe I will create a ticket for that). Since I think we don't have a strong/explicit relationship between  `numberOfRequestedOverdraftMemorySegments`, `numberOfRequestedMemorySegments`, `availableMemorySegment` and `currentPoolSize` for example:
- `!availableMemorySegment.isEmpty() && numberOfRequestedOverdraftMemorySegments != 0` - never should be true
- `numberOfRequestedMemorySegments == currentPoolSize && availableMemorySegment.isEmpty()` - always should be true when `numberOfRequestedOverdraftMemorySegments > 0`
- etc.

Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java:515]: Thanks @akalash for the timely feedback.

It makes sense to remove `numberOfRequestedOverdraftMemorySegments == 0` in another PR. Have to say, I like your suggestion about refactoring this class as the current logic looks bloated and a little confusing. Once we can sort out the relationship between these concepts, it can also help us understand whether we need to keep `numberOfRequestedOverdraftMemorySegments == 0`. 

I will keep this condition in this PR first and take a closer look at whether it's necessary to remove it when the refactoring is complete.

Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java]: Hi @akalash @reswqa , thanks for your quick feedback. Sounds make sense.

@reswqa could you squash the last commit to the [FLINK-31293] commit?

Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java]: @1996fanrui Sure, the reason why I use fix-up is to make it easier for everyone to see the changes I made last time. If there are no other issues, I'll squash it.

**Review by 1996fanrui — APPROVED** [review]
Thanks for the fix, LGTM.

And please wait for @akalash 's approval.

**Review by akalash — APPROVED** [review]
Thanks, good job. LGTM. Just a small question. Is there an easy way to check that we allocate an overdraft buffer only when isRequestedSizeReached? I mean we can create  test version of networkBufferPool  to check this but it doesn't look good/easy

**Comment by reswqa** [comment]
Thanks @akalash and @1996fanrui for the review.

> Is there an easy way to check that we allocate an overdraft buffer only when isRequestedSizeReached? I mean we can create test version of networkBufferPool to check this

The main reasons why I don't think there's an easy way are:
- `NetworkBufferPool` does not distinguish between `overdraft` and `non-overdraft` buffers request.
-  `NetworkBufferPool` itself is not an interface, it is not easy to make a very beautiful mock for it (i.e. `TestingNetworkBufferPool`), of course it can also be done through inheritance, but this will make our test still partly dependent on its implementation.

One solution I can think of is: we keep requesting and returning buffers until there are `currentPoolSize` buffers in `availableMemorySegment`. Next, we intercept the `requestPooledMemorySegment` method of the test version 
 of `NetworkBufferPool` and keep requesting buffers. Then we check that this method will be called only `isRequestedSizeReached()` is satisfied. Is this really necessary compared to the complexity? The tests introduced in this PR already cover this part, at least to some extent.

Perhaps it is enough if we can adding more checks for tests like `testDecreasePoolSize` to ensure that the condition must be met before requesting the overdraft buffer. WDYT?

**Comment by reswqa** [comment]
FYI：If there are no other concerns, I will merge this PR tomorrow.

**Comment by 1996fanrui** [comment]
> FYI：If there are no other concerns, I will merge this PR tomorrow.

Thanks for your work, it's ok from my side.

**Comment by akalash** [comment]
it is ok from me as well but doesn't forget to squash the two last commits

**Comment by reswqa** [comment]
Squash the two last commits. Merged % CI green.

**Pull request closed without merging**
